### PR TITLE
Lock records for editing via a plugin

### PR DESCRIPTION
**Pull request by ptn**
Issue #35

WIP, just getting things moving here.

What seemed more gemifiable was to use a Rails plugin that adds a `acts_as_lockable` method to ActiveRecord::Base.

Sample usage:

``` ruby
class MyModel < ActiveRecord::Base
  acts_as_lockable  # The lockable field will be :body, you can override with the :field option
end


MyModel.lockable_field # => :body
mymodel = MyModel.new
mymodel.lock(userA)
mymodel.locked? # => true
mymodel.has_lock?(userB) # => false
mymodel.save # Validation error: no user passed, we need to compare with the user that owns the lock
mymodel.save(userB) # Validation error: userB doesn't own the lock
mymodel.save(userA) # => true
mymodel.unlock(userB) # => false
mymodel.unlock(userA) # => true
mymodel.locked? # => false
mymodel.save # No need to pass a user anymore
```

The basic idea is what @jordanbyron suggested: simply use a validator to disallow saving when the user trying to do the save is different from the one that owns the lock. Since `save` doesn't usually record the user that tried to do the save, I had to add an instance variable and override `save` to set that. I think everyone would prefer it if we avoided `class_eval` and friends, so is there another way to do the override?

Next steps:
1. Write tests for the plugin
2. You know...actually use it in a model :wink:

Other things to take care of, at a much later date:
1. Create the Lock model when installing the plugin
2. Provide a generator that creates the migrations that `acts_as_lockable` requires when used in a model.

### CODE REVIEW

**Comment by ptn** [comment]
> I think everyone would prefer it if we avoided class_eval and friends, so is there another way to do the override?

Disregard this. abae963 fixes a different bug and also took care of this.

**Comment by ptn** [comment]
> \2. Provide a generator that creates the migrations that acts_as_lockable requires when used in a model.

Also forget about this. This was from a previous idea where the Lock model didn't exist yet.

Locking is now working on the shell. I added it to Activity, Article and Page so others can try it too.

File: vendor/plugins/acts_as_lockable/lib/acts_as_lockable.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  module ClassMethods
+    def acts_as_lockable(opts={})
+      include LockableMethods
+
+      attr_reader :tried_to_save
+
+      field = opts.fetch(:field, :body)
+      cattr_accessor :lockable_field
+      self.lockable_field = field
```
Comment (ptn) [inline vendor/plugins/acts_as_lockable/lib/acts_as_lockable.rb]: @jordanbyron do you think it's is necessary to query the lockable field from outside of the class? If not, I'd rather remove these lines, but if we need it, then I think it's best to make it readonly.

**Comment by jordanbyron** [comment]
Hey Pablo,

Sorry I haven't responded sooner. Your pull request was the last to come in and I've been loosely operating on a FIFO priority queue :wink2:

Anyway, it looks like you've gone ahead and made this a gem from the get go. That's great, but it means we need to have some more in-depth conversations about your code. I think it's OK to develop this gem in community, but as soon as it is ready I'd like to pull it out and release it properly. 

Naming gems with "ActsAs" is a little dated and I think we can do better here. I checked rubygems and there isn't a gem called `lockable` and I think that name is much better. Then instead of calling `acts_as_lockable` inside the model do `include Lockable` which is more in line with idiomatic Ruby code. How do you feel about that? 

Another thing, since this is a proper gem, you'll want to move the `Lock` model and migration into the gem itself. Otherwise we'll never been able to easily use this code in another application. 

So let's talk about how the `save` method works. I think it would make more sense to include an additional attribute on lockable models, say for example `saved_by`, which is required and is checked when the lock validations run. If we did that then the `update_attributes` method will still work and we don't have to change the signature of `save`.

I think that's enough to keep you busy for now. Please let me know if any of this is unclear or just plain wrong :smile: Thanks for working on this Pablo!

**Comment by ptn** [comment]
> I checked rubygems and there isn't a gem called lockable and I think that name is much better. Then instead of calling acts_as_lockable inside the model do include Lockable which is more in line with idiomatic Ruby code. How do you feel about that?

Yes! I like that name a lot more, and I feel that the architecture of a Rails plugin as taught in the docs has a ton of unnecessary abstraction around it. Let's change anything that makes things more direct and to the point. (BTW, did you ever stumble upon [this article](http://yehudakatz.com/2009/11/12/better-ruby-idioms/) by Yehuda? It discusses this exact point.)

Our new `Lockable` module would need to add a private class method to identify the lockable field:

``` ruby
class Article
  include Lockable
  # This method is added by Lockable; if a model doesn't call it, :body is assumed.
  lockable_field :text
end
```

> Another thing, since this is a proper gem, you'll want to move the Lock model and migration into the gem itself. Otherwise we'll never been able to easily use this code in another application.

So the plugin's `install.rb` file would copy them over to the appropriate directories, right?

> I think it would make more sense to include an additional attribute on lockable models, say for example saved_by, which is required and is checked when the lock validations run.

Commit 11a75c9 implements this. Here's a sample usage:

``` ruby

class Article
  include Lockable
end

article = Article.first
article.acquire_lock(current_user)
article.body = "This gem is awesome"
article.saved_by = current_user
article.save
article.release_lock(current_user)
```

I'm using a virtual attribute, is this what you had in mind or would you like to persist `saved_by` to the database? This sounds a bit like auditing.

> Thanks for working on this Pablo!

LET'S DO THIS!!!1

**Comment by jordanbyron** [comment]
Closing so we can take a new approach

**Pull request closed without merging**